components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java

```diff
@@ -697,18 +697,23 @@ private String handleUserConsent(HttpServletRequest request, String consent, OAu
             return EndpointUtil.getErrorRedirectURL(oauthProblemException, oauth2Params);
         }
 
-        //When responseType equal to "id_token" the resulting token is passed back as a query parameter
-        //According to the specification it should pass as URL Fragment
-        if (OAuthConstants.ID_TOKEN.equalsIgnoreCase(responseType)) {
-            if (authzRespDTO.getCallbackURI().contains("?")) {
-                return authzRespDTO.getCallbackURI() + "#" + oauthResponse.getLocationUri().substring(
-                        authzRespDTO.getCallbackURI().length() + 1, oauthResponse.getLocationUri().length());
+        //When response_mode equals to form_post, body parameter is passed back.
+        if (RESPONSE_MODE_FORM_POST.equals(oauth2Params.getResponseMode())
+                && StringUtils.isNotEmpty(oauthResponse.getBody())) {
+            return oauthResponse.getBody();
+        } else {
+            //When responseType equal to "id_token" the resulting token is passed back as a query parameter
+            //According to the specification it should pass as URL Fragment
+            if (OAuthConstants.ID_TOKEN.equalsIgnoreCase(responseType)) {
+                if (authzRespDTO.getCallbackURI().contains("?")) {
+                    return authzRespDTO.getCallbackURI() + "#" + StringUtils.substring(oauthResponse.getLocationUri()
+                            , authzRespDTO.getCallbackURI().length() + 1);
+                } else {
```

Comment:
Just a small clarification, can oauthResponse.getLocationUri() be null under any condition when it comes here?

Reply:
No. when creating the oauthResponse it will throw an exception if location is null.

Reply:
We can use guard clause here since there is nest of if-else condition block here.
https://medium.com/better-programming/why-you-need-to-stop-using-else-statements-5b1fd09dea9e 

ex:
if (! OAuthConstants.ID_TOKEN.equalsIgnoreCase(responseType)) {
return oauthResponse.getLocationUri().replace("?", "#");
}
if(! authzRespDTO.getCallbackURI().contains("?")){
return oauthResponse.getLocationUri().replace("?", "#");
}
                    return authzRespDTO.getCallbackURI() + "#" + StringUtils.substring(oauthResponse.getLocationUri()
                            , authzRespDTO.getCallbackURI().length() + 1);